canonical/istio-operators
# feat: enable certificates relation on istio-pilot

Enabling the certificates relation enables the integration with tls-certificates-operator and other charms that provide CA certificates. These CA certificates are used by istio-pilot to configure TLS on the ingress Gateway resource.
In the past this was done exclusively via charm configuration, which is now removed in favour of CA certificates provided through the certificates relation.

#### Manual testing
1. Deploy the following:

```bash
juju deploy istio-gateway istio-ingressgateway --trust --config kind=ingress --channel latest/edge
juju deploy self-signed-certificates --channel edge
```
2. In parallel, build `istio-pilot` from this PR's branch
3. Deploy `istio-pilot` and relate
```bash
juju deploy ./istio-pilot*.charm --trust
juju relate istio-pilot istio-ingressgateway
juju relate istio-pilot:certificates self-signed-certificates:certificates
```
5. Get the Gateway (`kubectl get Gateway <name of default-gateway> -n<namespace> -oyaml`) and Secret (secret name should be <default-gateway name>-gateway-secret) to see the certificate in place. The Gateway should show this:
```yaml
    port:
      name: https
      number: 8443
      protocol: HTTPS
    tls:
      mode: SIMPLE
```
The Secret should have a cert and key in a long encoded string.

7. Additionally you can curl the ingressgateway (curl -v IP.nio.io usually) to see TLS in action

TODO:
- [x] Increase test coverage
- [x] Fix linting issues
- [x] Provide manual testing instructions
